File: src/15.三数之和.rs
/*
 * @lc app=leetcode.cn id=15 lang=rust
 *
 * [15] 三数之和
 */
struct Solution;
// @lc code=start
use std::cmp::Ordering;
impl Solution {
    #[inline]
    fn two_sum_sorted(nums: &[i32], target: i32) -> Vec<Vec<i32>> {
        let n = nums.len();
        let (mut left, mut right) = (0, n - 1);
        let mut ans = Vec::new();
        while left < right {
            let (a, b) = (nums[left], nums[right]);
            let sum = a + b;
            if sum == target {
                ans.push(vec![nums[left], nums[right]]);
            }
            match sum.cmp(&target) {
                Ordering::Equal | Ordering::Less => {
                    // move left
                    while nums[left] == a && left < right {
                        left += 1;
                    }
                }
                Ordering::Greater => {
                    // move right
                    while nums[right] == b && left < right {
                        right -= 1;
                    }
                }
            }
        }
        ans
    }
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();
        if n < 3 {
            return vec![];
        }
        nums.sort_unstable();
        let mut ans = vec![];
        let mut i = 0;
        while i < n - 2 {
            let first = nums[i];
            if first > 0 {
                break;
            }
            for mut rest in Self::two_sum_sorted(&nums[i + 1..], -first) {
                rest.insert(0, first);
                ans.push(rest);
            }
            while nums[i] == first && i < n - 2 {
                i += 1;
            }
        }
        ans
    }
}
// @lc code=end
mod utils;
#[test]
fn test_solution() {
    macro_rules! test {
        ($n:tt, $ans:tt) => {
            let ans: Vec<Vec<i32>> = vec2d!$ans;
            assert_eq!(
                Solution::three_sum(vec!$n),
                ans
            );
        }
    };
    test!([-1, 0, 1, 2, -1, -4], [[-1, -1, 2], [-1, 0, 1]]);
    test!([1, 1, 1], []);
    test!([-1, 0, 1], [[-1, 0, 1]]);
    test!([], []);
    test!([0, 0, 0, 0], [[0, 0, 0]]);
    test!([0, 1, 1], []);
}

File: src/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from prelude import *
# @lc code=start


class Solution:
    def twoSum(self, nums: List[int], i: int, j: int, target: int):
        l, r = i, j - 1
        ans = []
        while l < r:
            left_val = nums[l]
            right_val = nums[r]
            sum = left_val + right_val
            if sum == target:
                ans.append([left_val, right_val])
                while l < r and nums[l] == left_val:
                    l += 1
            elif sum < target:
                while l < r and nums[l] == left_val:
                    l += 1
            else:
                while l < r and nums[r] == right_val:
                    r -= 1
        return ans

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        ans = []
        i = 0
        while i < n:
            first = nums[i]
            for second, third in self.twoSum(nums, i+1, n, target=-first):
                ans.append([first, second, third])
            while i < n and nums[i] == first:
                i += 1
        return ans


# @lc code=end
if __name__ == '__main__':
    def test(input, expected):
        calc = Solution().threeSum(input)
        calc = set(tuple(sorted(i)) for i in calc)
        expected = set(tuple(sorted(i)) for i in expected)
        if calc != expected:
            print(f'case failed: `{input}`')
            print(f'  calc     = `{calc}`')
            print(f'  expected = `{expected}`')
            exit(1)
    test(
        [-1, 0, 1, 2, -1, -4],
        [
            [-1, 0, 1],
            [-1, -1, 2]
        ]
    )
